### PR TITLE
DDP-6884 extend address component

### PR DESCRIPTION
**Pull request by aglushko-broad**

### CODE REVIEW

**Comment by MocanaAtBroad** [comment]
I am going to request that you write me out the story of what you are thinking of doing. Not saying there is anything wrong with code, but please walk me through it as I am reviewing A LOT of PRs these days with very limited time and a few comments can really help all reviewers get through it that much faster.
Comments in code are cool too. In many cases, preferable.
(I also find that when I add comments it clarifies my own thinking of what I am doing).
I'll be making same request from everybody else and I'll be trying to do same for my own PRs.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 1]
```diff
@@ -42,22 +42,31 @@ import { AddressError } from '../../models/addressError';
...
 import { ConfigurationService } from '../../services/configuration.service';
 
+interface FormError {
+    message: string;
+    isEasyPostError: boolean;
+}
+
+interface FieldError extends AddressError {
+    isEasyPostError: boolean;
+}
```
Comment (aglushko-broad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:58]: extend all types of errors with `isEasyPostError` flag in order to divide errors and calculate `isEasyPostInvalid` value

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 2]
```diff
@@ -189,17 +201,19 @@ export class AddressEmbeddedComponent implements OnDestroy, OnInit {
...
     public errorMessagesToDisplay$: Observable<string[]>;
     public suggestionForm: FormGroup;
+    public ignoreEasyPostErrorsCheckbox = new FormControl();
     public isInputComponentBusy$ = new BehaviorSubject<boolean>(false);
     // variables for template
     public suggestionInfo$: Observable<AddressSuggestion | null>;
-    public inputAddress$: Subject<Address | null> = new BehaviorSubject(null);
-    public verifyFieldErrors$: Observable<AddressError[]>;
+    public inputAddress$ = new BehaviorSubject<Address | null>(null);
+    public verifyFieldErrors$: Observable<FieldError[]>;
```
Comment (aglushko-broad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:213]: just refactoring for these fields

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 3]
```diff
@@ -189,17 +201,19 @@ export class AddressEmbeddedComponent implements OnDestroy, OnInit {
...
     public isReadOnly$: Observable<boolean>;
-    public inputComponentAddress$: Subject<Address | null> = new BehaviorSubject(null);
-    public formValidStatusChanged$: Subject<boolean> = new BehaviorSubject(true);
+    public inputComponentAddress$ = new BehaviorSubject<Address | null>(null);
+    public formValidStatusChanged$ = new BehaviorSubject(true);
```
Comment (aglushko-broad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:216]: just refactoring for these fields

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 4]
```diff
@@ -322,8 +337,11 @@ export class AddressEmbeddedComponent implements OnDestroy, OnInit {
                 } : null, {onlyDistinct: false})
         );
 
-        this.errorMessagesToDisplay$ = this.state$.pipe(
-            extract(state => state.formErrorMessages.concat(state.formWarningMessages)));
+        this.errorMessagesToDisplay$ = combineLatest([this.state$, ignoreEasyPostErrorsObservable]).pipe(
+            extract(([state, ignoreEasyPost]) => state.formErrorMessages
+                .filter(error => !(ignoreEasyPost && error.isEasyPostError))
```
Comment (aglushko-broad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:343]: do not display EasyPost errors if `ignoreEasyPost` checkbox is checked

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 5]
```diff
@@ -350,7 +368,7 @@ export class AddressEmbeddedComponent implements OnDestroy, OnInit {
                 }
             }),
             tap((error: string | null) => {
-                this.stateUpdates$.next({formErrorMessages: error ? [error] : []});
+                this.stateUpdates$.next({formErrorMessages: error ? [{ message: error, isEasyPostError: false }] : []});
```
Comment (aglushko-broad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:372]: any the angular form state validation error is not an easy post error

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 6]
```diff
@@ -395,15 +413,24 @@ export class AddressEmbeddedComponent implements OnDestroy, OnInit {
...
+                        showSuggestion: true
                     });
                 } else {
                     this.stateUpdates$.next({
-                        formErrorMessages: [], formWarningMessages: enteredWarningMessages,
-                        warnings: addressSuggestion.warnings, suggestedAddress: null, enteredAddress: null, showSuggestion: false
+                        ...commonStateFields,
+                        suggestedAddress: null,
+                        enteredAddress: null,
+                        showSuggestion: false
```
Comment (aglushko-broad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:434]: only refactoring from 416 line

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 7]
```diff
@@ -487,13 +514,9 @@ export class AddressEmbeddedComponent implements OnDestroy, OnInit {
...
                         // These are the "global" errors reported by EasyPost
-                        errorUpdate.formErrorMessages.push(currError.message);
-                    } else if (isStreetRequiredError(currError)) {
-                        // Seems like EasyPost needs a street address before it verifies other fields.
-                        // Since street1 might not be filled yet, transform message into a "global" error message.
-                        errorUpdate.formErrorMessages.push('Street address is missing, could not verify address.');
+                        errorUpdate.formErrorMessages.push({ message: currError.message, isEasyPostError: true });
                     } else {
-                        errorUpdate.fieldErrors.push(currError);
+                        errorUpdate.fieldErrors.push({ ...currError, isEasyPostError: true });
```
Comment (aglushko-broad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:520]: verificationError$ stream contains only easyPost errors

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 8]
```diff
@@ -519,19 +542,27 @@ export class AddressEmbeddedComponent implements OnDestroy, OnInit {
 
         // "Real" as opposed to "Temp". Important we return the saved address as properties are added on server
         const saveRealAddressAction$ = this.saveTrigger$.pipe(
-            withLatestFrom(this.formErrorMessages$, currentAddress$),
-            filter(([_, formErrors, addressToSave]) => {
-                return !formErrors?.length && canSaveRealAddress(addressToSave);
+            withLatestFrom(currentAddress$, this.formErrorMessages$, this.verifyFieldErrors$, ignoreEasyPostErrorsObservable),
+            filter(([_, addressToSave, formErrors, fieldErrors, ignoreEasyPost]) => {
+                return !formErrors?.filter(error => !(ignoreEasyPost && error.isEasyPostError)).length
```
Comment (aglushko-broad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:548]: allow to save address if ignoreEasyPost is checked and all errors are easy post errors

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 9]
```diff
@@ -519,19 +542,27 @@ export class AddressEmbeddedComponent implements OnDestroy, OnInit {
...
+                    && !fieldErrors?.filter(error => !(ignoreEasyPost && error.isEasyPostError)).length
```
Comment (aglushko-broad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:549]: do not allow to save address if there are field errors as well

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 10]
```diff
@@ -519,19 +542,27 @@ export class AddressEmbeddedComponent implements OnDestroy, OnInit {
...
+                    && canSaveRealAddress(addressToSave);
             }),
             tap(() => busyCounter$.next(1)),
-            concatMap(([_, __, addressToSave]) => this.addressService.saveAddress(addressToSave, false)),
+            concatMap(([_, addressToSave]) => this.addressService.saveAddress(addressToSave, false)),
+            catchError((error) => {
```
Comment (aglushko-broad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:554]: I ran into errors for saving address couple of times, so there is a sence to add catchError for that.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 11]
```diff
@@ -519,19 +542,27 @@ export class AddressEmbeddedComponent implements OnDestroy, OnInit {
...
+                return of(null);
+            }),
             tap(() => busyCounter$.next(-1)),
             share()
         );
 
         const deleteTempAddressAction$ = saveRealAddressAction$.pipe(
             withLatestFrom(this.state$),
-            filter(([_, state]) => !!state.activityInstanceGuid),
+            filter(([address, state]) => !!address && !!state.activityInstanceGuid),
```
Comment (aglushko-broad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:566]: do not delete temp address if saving address failed

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 12]
```diff
@@ -582,15 +613,28 @@ export class AddressEmbeddedComponent implements OnDestroy, OnInit {
...
-            activityRequirementsMet$
+            activityRequirementsMet$,
+            ignoreEasyPostErrorsObservable
         ]).pipe(
-            map(([formValidStatusChanged, formErrors, addressErrors, reqsMet]) =>
+            map(([formValidStatusChanged, formErrors, addressErrors, reqsMet, ignoreEasyPost]) =>
                 (!this.configuration.addressEnforceRequiredFields || formValidStatusChanged)
-                && !formErrors.length && !addressErrors.length && reqsMet),
+                && !formErrors.filter(error => !(ignoreEasyPost && error.isEasyPostError)).length
+                && !addressErrors.filter(error => !(ignoreEasyPost && error.isEasyPostError)).length
```
Comment (aglushko-broad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:623]: update validStatusChanged depending on ignoreEasyPost  checkbox state

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 13]
```diff
@@ -582,15 +613,28 @@ export class AddressEmbeddedComponent implements OnDestroy, OnInit {
...
 
+        combineLatest([
+            this.formErrorMessages$,
+            this.verifyFieldErrors$,
+        ]).pipe(
+            map(([formErrors, fieldErrors]) =>
+                (formErrors.length || fieldErrors.length)
+                && formErrors.every(error => error.isEasyPostError)
+                && fieldErrors.every(error => error.isEasyPostError)),
+        ).subscribe(this.isEasyPostInvalid$);
```
Comment (aglushko-broad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:637]: set isEasyPostInvalid state in order to show/hide checkbox

**Comment by aglushko-broad** [comment]
> I am going to request that you write me out the story of what you are thinking of doing. Not saying there is anything wrong with code, but please walk me through it as I am reviewing A LOT of PRs these days with very limited time and a few comments can really help all reviewers get through it that much faster.
> Comments in code are cool too. In many cases, preferable.
> (I also find that when I add comments it clarifies my own thinking of what I am doing).
> I'll be making same request from everybody else and I'll be trying to do same for my own PRs.

agree with you! done! hope it's better now :)

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 14]
```diff
@@ -106,7 +115,10 @@ interface AddressSuggestion {
                     </mat-radio-group>
                 </mat-card-content>
             </mat-card>
-        </form>`,
+        </form>
+         <mat-checkbox *ngIf="isEasyPostInvalid$ | async" [formControl]="ignoreEasyPostErrorsCheckbox" class="ignore-easy-post-errors">
+             Please use the address as entered
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts]: Let i18n this text.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 15]
```diff
@@ -189,17 +201,19 @@ export class AddressEmbeddedComponent implements OnDestroy, OnInit {
...
+    public formErrorMessages$: Observable<FormError[]>;
     public errorMessagesToDisplay$: Observable<string[]>;
     public suggestionForm: FormGroup;
+    public ignoreEasyPostErrorsCheckbox = new FormControl();
     public isInputComponentBusy$ = new BehaviorSubject<boolean>(false);
     // variables for template
     public suggestionInfo$: Observable<AddressSuggestion | null>;
-    public inputAddress$: Subject<Address | null> = new BehaviorSubject(null);
-    public verifyFieldErrors$: Observable<AddressError[]>;
+    public inputAddress$ = new BehaviorSubject<Address | null>(null);
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts]: Your code here is shorter, but the `Subject` type was there on purpose; don't want it used as BehaviorSubject except for initialization. BehaviorSubject allows for value to be set. We don't want within code.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts (hunk 1 above)
Comment (MocanaAtBroad) [reply ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:58]: Suggestion: create IsEasyPostError interface and  have form error and FieldError extend from it so you can use them uniformly, e.g.:
```typescript
interface IsEasyPostError {
    isEasyPostError: boolean;
}

interface FormError extends IsEasyPostError {
    message: string;
}

export interface FieldError extends AddressError, IsEasyPostError {}

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts (hunk 2 above)
Comment (MocanaAtBroad) [reply ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:213]: I think can stay as `AddressError[]`

**Review by MocanaAtBroad — COMMENTED** [review]
It looks good! I have a couple of refactoring comments, particularly it occurs to me that validation that allows saving address ought to be the same used to emit value to indicate whether component has a valid address or not. But take a look.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts (hunk 15 above)
Comment (aglushko-broad) [reply ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts]: Actually you just described what's BehaviorSubject is. It's a Subject but with initial value. Subject also allows to set a value (subject.next(...))

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts (hunk 2 above)
Comment (aglushko-broad) [reply ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:213]: well, fields errors may contain easyPost error, and we use this data. So we need FieldError interface here (with isEasyPostError property)

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts (hunk 1 above)
Comment (aglushko-broad) [reply ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:58]: good idea, done!

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts (hunk 14 above)
Comment (aglushko-broad) [reply ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts]: thanks! done

File: ddp-workspace/projects/ddp-pancan/src/assets/i18n/en.json [hunk 16]
```diff
@@ -757,7 +757,8 @@
                 "E.TIMED_OUT": "Exceeded max timeout.",
                 "E.TIME_ZONE.UNAVAILABLE": "The time zone service is currently unavailable.",
                 "E.PO_BOX.INTERNATIONAL": "Cannot verify international PO Box. Please note, this can show up as an error on a successful request."
-            }
+            },
+            "UseAsEntered":  "Please use the address as entered"
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-pancan/src/assets/i18n/en.json]: We are unable to confirm that this address is valid. Check this box to use this address entered.

File: ddp-workspace/projects/ddp-pancan/src/assets/i18n/es.json [hunk 17]
```diff
@@ -755,7 +755,8 @@
                 "E.TIMED_OUT": "Se superó el tiempo máximo de espera.",
                 "E.TIME_ZONE.UNAVAILABLE": "El servicio de zona horaria no está disponible en este momento.",
                 "E.PO_BOX.INTERNATIONAL": "No se puede verificar la casilla postal internacional. Tenga en cuenta que eso puede causar errores en solicitudes correctas."
-            }
+            },
+            "UseAsEntered":  "[ES]Please use the address as entered"
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-pancan/src/assets/i18n/es.json]: No podemos confirmar que esta dirección es válida. Marque esta casilla para confirmar el uso de esta dirección como ha sido ingresada.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts (hunk 15 above)
Comment (MocanaAtBroad) [reply ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts]: I worry about the usage of getValue().
https://rxjs.dev/api/index/class/BehaviorSubject#getvalue-

Really want it to be used as a Subject in body of code

File: ddp-workspace/projects/ddp-pancan/src/assets/i18n/es.json (hunk 17 above)
Comment (aglushko-broad) [reply ddp-workspace/projects/ddp-pancan/src/assets/i18n/es.json]: thanks, done!

File: ddp-workspace/projects/ddp-pancan/src/assets/i18n/en.json (hunk 16 above)
Comment (aglushko-broad) [reply ddp-workspace/projects/ddp-pancan/src/assets/i18n/en.json]: thanks, done!

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts (hunk 15 above)
Comment (aglushko-broad) [reply ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts]: sure! done

File: ddp-workspace/projects/ddp-pancan/src/assets/i18n/en.json (hunk 16 above)
Comment (MocanaAtBroad) [reply ddp-workspace/projects/ddp-pancan/src/assets/i18n/en.json]: Sorry! I screwed up with my grammar!
"We are unable to confirm that the address is valid. Please check this box to use the address as entered"

Comment (aglushko-broad) [reply ddp-workspace/projects/ddp-pancan/src/assets/i18n/en.json]: done!

**Review by MocanaAtBroad — APPROVED** [review]

**Pull request merged**